Source/SourceKittenFramework/Request.swift

```diff
@@ -103,16 +104,59 @@ Cache SourceKit requests for strings from UIDs
 internal func stringForSourceKitUID(uid: sourcekitd_uid_t) -> String? {
     if let string = uidStringMap[uid] {
         return string
-    } else if let uidString = String(UTF8String: sourcekitd_uid_get_string_ptr(uid)) {
-        let uidString = SwiftDocKey(rawValue: uidString)?.rawValue ??
-            SwiftDeclarationKind(rawValue: uidString)?.rawValue ??
-            SyntaxKind(rawValue: uidString)?.rawValue ?? uidString
+    }
+    let length = sourcekitd_uid_get_length(uid)
+    let pointer = UnsafeMutablePointer<Int8>(sourcekitd_uid_get_string_ptr(uid))
+    if let uidString = String(bytesNoCopy: pointer, length: length,
+        encoding: NSUTF8StringEncoding, freeWhenDone: false) {
+        /*
+        `String` created by `String(UTF8String:)` is based on `NSString`.
+        `NSString` base `String` has performance penalty on getting `hashValue`.
+        Everytime on getting `hashValue`, it calls `decomposedStringWithCanonicalMapping` for
+        "Unicode Normalization Form D" and creates autoreleased `CFString (mutable)` and
+        `CFString (store)`. Those `CFString` are created every time on using `hashValue`, such as
+        using `String` for Dictionary's key or adding to Set.
+        
+        For avoiding those penalty, replaces with enum's rawValue String if defined in SourceKitten.
+        That does not cause calling `decomposedStringWithCanonicalMapping`.
+        */
+        let uidString = sourceKittenRawValueStringFrom(uidString) ?? "\(uidString)"
         uidStringMap[uid] = uidString
         return uidString
     }
     return nil
 }
 
+/**
+Returns SourceKitten defined enum's rawValue String from string
+
+- parameter uidString: String created from sourcekitd_uid_get_string_ptr().
+
+- returns: rawValue String if defined in SourceKitten, nil otherwise.
+*/
+private func sourceKittenRawValueStringFrom(uidString: String) -> String? {
+    return SwiftDocKey(rawValue: uidString)?.rawValue ??
+        SwiftDeclarationKind(rawValue: uidString)?.rawValue ??
+        SyntaxKind(rawValue: uidString)?.rawValue
+}
+
+/**
+Returns Swift's native String from NSUTF8StringEncoding bytes and length
+
+`String(UTF8String:)` creates `String` based on `NSString`.
+That is slower than Swift's native String on some scene.
+ 
+- parameter bytes: UnsafePointer<Int8>
+- parameter length: length of bytes
+
+- returns: String Swift's native String
+*/
+private func swiftStringFrom(bytes: UnsafePointer<Int8>, length: Int) -> String? {
+    let pointer = UnsafeMutablePointer<Int8>(bytes)
+    return String(bytesNoCopy: pointer, length: length, encoding: NSUTF8StringEncoding,
+        freeWhenDone: false).map { "\($0)" }
```

Comment:
It's converting from the `NSString` backed swift string to a native swift string. The c'tor is actually `-[NSString initWithBytesNoCopy:length:encoding:freeWhenDone]`